pkg/controller/handler.go

```diff
@@ -181,39 +181,53 @@ func (p *Propeller) Handle(ctx context.Context, namespace, name string) error {
 		}
 	}
 
-	mutatedWf, err := p.TryMutateWorkflow(ctx, w)
-	if err != nil {
-		// NOTE We are overriding the deepcopy here, as we are essentially ingnoring all mutations
-		// We only want to increase failed attempts and discard any other partial changes to the CRD.
-		mutatedWf = RecordSystemError(w, err)
-		p.metrics.SystemError.Inc(ctx)
-	} else if mutatedWf == nil {
-		return nil
-	} else {
-		if !w.GetExecutionStatus().IsTerminated() {
-			// No updates in the status we detected, we will skip writing to KubeAPI
-			if mutatedWf.Status.Equals(&w.Status) {
-				logger.Info(ctx, "WF hasn't been updated in this round.")
-				return nil
+	for ; ; {
+		mutatedWf, err := p.TryMutateWorkflow(ctx, w)
+		if err != nil {
+			// NOTE We are overriding the deepcopy here, as we are essentially ingnoring all mutations
+			// We only want to increase failed attempts and discard any other partial changes to the CRD.
+			mutatedWf = RecordSystemError(w, err)
+			p.metrics.SystemError.Inc(ctx)
+		} else if mutatedWf == nil {
+			return nil
+		} else {
+			if !w.GetExecutionStatus().IsTerminated() {
+				// No updates in the status we detected, we will skip writing to KubeAPI
+				if mutatedWf.Status.Equals(&w.Status) {
+					logger.Info(ctx, "WF hasn't been updated in this round.")
+					return nil
+				}
+			}
+			if mutatedWf.GetExecutionStatus().IsTerminated() {
+				// If the end result is a terminated workflow, we remove the labels
+				// We add a completed label so that we can avoid polling for this workflow
+				SetCompletedLabel(mutatedWf, time.Now())
+				ResetFinalizers(mutatedWf)
 			}
 		}
-		if mutatedWf.GetExecutionStatus().IsTerminated() {
-			// If the end result is a terminated workflow, we remove the labels
-			// We add a completed label so that we can avoid polling for this workflow
-			SetCompletedLabel(mutatedWf, time.Now())
-			ResetFinalizers(mutatedWf)
+		// TODO we will need to call updatestatus when it is supported. But to preserve metadata like (label/finalizer) we will need to use update
+
+		// update the GetExecutionStatus block of the FlyteWorkflow resource. UpdateStatus will not
+		// allow changes to the Spec of the resource, which is ideal for ensuring
+		// nothing other than resource status has been updated.
+		newWf, updateErr := p.wfStore.Update(ctx, mutatedWf, workflowstore.PriorityClassCritical)
```

Comment:
Time to switch to UpdateStatus?

Reply:
I want to do that as a follow up. It is a little dangerous to do that at the moment, because we add a finalizer in some places. Do not want to de-stabilize